### PR TITLE
Fix for issue3103 (race condition detection)

### PR DESCRIPTION
**Pull request by DrTodd13**
Okay.  Second try looks better.

### CODE REVIEW

**Comment by seibert** [comment]
This seems to be breaking a number of parfor tests, for example:

https://travis-ci.org/numba/numba/jobs/405977860#L3509-L3597

**Comment by DrTodd13** [comment]
Okay.  Thanks for the report.  Just pushed a fix.

**Comment by stuartarchibald** [comment]
Thanks for the patch. This approach seems to solve the reported problem. I wonder a couple of things:
  1. If a race condition is detected, should a warning be raised? Are there any valid uses for such a race?
  2. If code like:
  ```python
    @skip_unsupported
    def test_parfor_race_2(self):
        def test_impl(x, y):
            for j in range(y):
                k = j
            return k
        self.prange_tester(test_impl, 10, 20)
  ```
  is written, it also is a race, but the reduce-and-select-first method won't work (assuming I've read the code correctly). Should there be a check on the reduction array to make sure that all the values are the same before returning one?

**Comment by DrTodd13** [comment]
The approach is if a race is detected, to turn the racing variable into a one-element array that is passed to all workers.  Then, the workers are all going to ping-pong the cache line around with the variable in it as they are assigning to it.  If the same value always gets assigned then you'll get the same answer as single-threaded.  If different values are set then you'll get the last value set by some worker in its iteration space.  It isn't like we remember the value set in each iteration and then make sure they are all the same before returning it.  Hey, they are writing invalid code so they are going to reap the performance penalty of the cache line ping-ponging and the undefinedness of all other race conditions.

Let me know how you want to handle the warning.  We should add it after the spot where we remove the reduction vars from the set of races to get the true set of races.

**Comment by stuartarchibald** [comment]
Thanks for explaining the approach, v. helpful. I think to be in line with other places in the compiler pipeline a `warnings.warn_explicit` should be used and the warning category should be either a `UserWarning` or perhaps inherit from that into a `numba.errors.ParallelSafetyWarning` or something so we could configure switching it off at some later point? I think the code should warn on absolutely any sort of detected race condition, irrespective of whether something can be done to reasonably handle it. Thanks.

**Comment by DrTodd13** [comment]
I added something but it looks like warn_explicit only prints anything in NUMBA_WARNINGS is on.  Is this the desired behavior because I imagine most people don't use that?

File: numba/npyufunc/parfor.py [hunk 1]
```diff
@@ -359,6 +378,16 @@ def _create_gufunc_for_parfor_body(
             set(parfor_outputs) -
             set(parfor_redvars)))
 
+    races = races.difference(set(parfor_redvars))
+    for race in races:
+        warnings.warn_explicit("Variable %s used in parallel loop may be written "
+                      "to simulatenously by multiple workers and may result "
```
Comment (stuartarchibald) [inline numba/npyufunc/parfor.py]: s/simulatenously/simultaneously/

**Comment by stuartarchibald** [comment]
I think because the `ParallelSafetyWarning` inherited from a `NumbaWarning` instead of a [`UserWarning`](https://docs.python.org/3/library/exceptions.html#UserWarning):
https://github.com/numba/numba/pull/3137/commits/2136fefa728a258e63338b8c16302c61acccbed3#diff-03285c9e2ae1987a63fa271f8d5792beR36
it will not appear unless `NUMBA_WARNINGS` is on. I think in this case, the the user needs to see this warning and so it needs to inherit from something not `NumbaWarning` related. The potential future ability to mask off this warning will also need to take a different path to `NUMBA_WARNINGS`.

It'd be great to augment the unit test(s) for this with a `warnings.catch_warnings()` to make sure that the warning is raised correctly too.

Thanks.

File: numba/npyufunc/parfor.py (hunk 1 above)
Comment (DrTodd13) [reply numba/npyufunc/parfor.py]: Done.

**Comment by stuartarchibald** [comment]
Thanks for the fixes. Merging.

**Pull request merged**